File: python/add_two_numbers.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None
    
    def __str__(self):
        '''converting the list to a string'''
        r = []
        while True:
            r.append(str(self.val))
            self = self.next
            if not self:
                break
        return '->'.join(r)

    def add_node(self, x):
        a = ListNode(x)
        #print a
        a.next = self
        self = a
        return self


class Solution:
    # @return a ListNode
    def addTwoNumbers(self, l1, l2):
        carry = 0
        r = []
        while l1 or l2:
            s = 0
            if l1:
                #print 'l1:'+str(l1.val)
                s += l1.val
                l1 = l1.next
            if l2:
                #print 'l2:'+str(l2.val)
                s += l2.val
                l2 = l2.next

            s += carry
            this_bit = s%10
            carry = s/10

            r.insert(0, this_bit)

        if carry > 0:
            r.insert(0, carry)

        r_link = None
        for i in range(len(r)):
            if i == 0:
                r_link = ListNode(r[i])
            else:
                tmp = ListNode(r[i])
                tmp.next = r_link
                r_link = tmp
        return r_link


if __name__ == '__main__':
    l1 = ListNode(5).add_node(4).add_node(3)
    l2 = ListNode(5).add_node(6).add_node(4)
    #l1 = ListNode(5)
    #l2 = ListNode(5)
    #l1 = ListNode(8).add_node(1)
    #l2 = ListNode(0)
    print l1, l2
    print Solution().addTwoNumbers(l1, l2)

File: c++/add_two_numbers.cpp
#include <iostream>
#include <vector>

using namespace std;

//Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};


ListNode* insert(ListNode* l, int x){
    ListNode *n = new ListNode(x);
    n->next = l;
    return n;
}


void print(ListNode* l){
    cout << l->val;
    while(l->next != NULL){
        cout << "->";
        l = l->next;
        cout << l->val;
    }
    cout << endl; 
}

class Solution {
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) {
        int carry = 0;
        ListNode* r = NULL;
        ListNode* re = NULL;
        while (l1 != NULL || l2 != NULL) {
            int this_digit = 0;
            if (l1 != NULL) {
                this_digit += l1->val;
                l1 = l1->next;
            }
            if (l2 != NULL) {
                this_digit += l2->val;
                l2 = l2->next;
            }
            int this_r = this_digit+carry;
            carry = this_r/10;
            ListNode* this_l = new ListNode(this_r%10);
            if (r == NULL) {
                r = this_l;
                re = r;
            } else {
                r->next = this_l;
                r = this_l;
            }
        }

        if (carry > 0) {
            ListNode* this_l = new ListNode(carry);
            if (r == NULL) {
                r = this_l;
                re = r;
            } else {
                r->next = this_l;
                r = this_l;
            } 
        }

        return re;
    }
};


int main(){
    Solution *s = new Solution();
    ListNode *l1 = new ListNode(5);
    l1 = insert(l1, 4);
    l1 = insert(l1, 3);
    print(l1);
    ListNode *l2 = new ListNode(5);
    l2 = insert(l2, 6);
    l2 = insert(l2, 4);
    print(l2);
    ListNode* r = s->addTwoNumbers(l1, l2);
    print(r);
    return 0;
}
